src/phase0/state_transition/mod.rs

```diff
@@ -70,6 +70,14 @@ pub enum InvalidOperation {
     IndexedAttestation(InvalidIndexedAttestation),
     #[error("invalid deposit: {0}")]
     Deposit(InvalidDeposit),
+    #[error("invalid block: {0}")]
+    BlockProcessing(InvalidBlock),
```

Comment:
while it may not exactly be called an "operation" in the spec, i think just having a variant for randao failures is preferred to keep the symmetry w/ the other errors here

and we can inline the one variant of randao processing w/ the invalid signature


```suggestion
    #[error("invalid randao (BLS signature): {0}")]
    Randao(BLSSignature),
```
i'll leave it up to you if you want to include more detailed error information in the enum variant, it could contain pubkey, message and signature but that could be overkill -- you pick

Reply:
Hope enum naming is alright, let me know if not: 62eaaca

Reply:
looks good!

although i meant that we can just use the actual signature data, not the most inner enum provided in 62eaaca

to do this we would delete the enum `BLSSignature` and adjust the `InvalidOperation::Randao` variant to just wrap a `crypto.Signature` (from the top-level module `crypto`)

```rust
// modify the import at the top:
use crate::crypto::{fast_aggregate_verify, hash, Signature as BLSSignature};

pub enum InvalidOperation {
    // other variants elided
    Randao(BLSSignature),
}
```

where i added the signature alias `BLSSignature` just to disambiguate from some other type of generic `Signature` that could be floating around

Reply:
and then at the call site something like:

```rust
        return Err(Error::InvalidOperation(InvalidOperation::Randao(
            body.randao_reveal.clone(),
        )));
```

then at the least a user of this library will see the faulty signature/reveal upon error

Reply:
Fixed: 1abdf5c